test/metabase/driver/h2_test.clj

```diff
@@ -207,3 +207,61 @@
                     {:database (u/the-id (mt/db))
                      :engine :h2
                      :native {:query trigger-creation-attempt}}))))))
+
+(deftest check-read-only-test
+  (testing "read only statements should pass"
+    (are [query] (nil?
+                  (#'h2/check-read-only-statements
+                   {:database (u/the-id (mt/db))
+                    :engine :h2
+                    :native {:query query}}))
+      "select * from orders"
+      "select 1; select 2;"
+      "explain select * from orders"
+      "values (1, 'Hello'), (2, 'World');"
+      "show tables"
+      "table orders"
+      "call 1 + 1"
+      ;; Note this passes the check, but will fail on execution
+      "update venues set name = 'bill'; some query that can't be parsed;"))
```

Comment:
Are we parsing the SQL, or are we asking H2 to parse it but not do anything?
What I'm driving at is, is there a possibility of drift between the parser responsible for `check-read-only-statements` and H2 itself?

Reply:
We use H2 classes to parse queries, but not execute them, and then inspect the results.

Reply:
Ah, makes sense. Then this seems sound to me.